In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
%store -r ted_2

In [5]:
# We checked that it came intact and now we will reassign it back to the original variableted 
ted = ted_2

In [6]:
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

   ...                    speaker_occupation  \
0  ...                       Author/educator   
1  ...                      Climate advocate   
2  ...                  Technology columnist   
3  ...    Activist for environmental justice   
4  ...  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  \
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520   
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292   
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550   
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869   

  comments_per_view  views_per_comment  film_datetime  \
0          0.000096       10372.745443     2006-02-25   
1          0.000083       12077.433962     2006-02-25   
2          0.000076       13195.903226     2006-02-24   
3          0.000118        8487.750000     2006-02-26   
4          0.000049       20245.984823     2006-02-22   

                                        ratings_list num_ratings  
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...       93850  
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...        2936  
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...        2824  
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...        3728  
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...       25620  

[5 rows x 22 columns]

# 8. Which occupations deliver the funniest TED talks on average?
Bonus:
- for each talk, calculate the most frequent rating
- for each talk, clean the occupation data so that there's only one occupation per talk

In [135]:
ted.speaker_occupation.value_counts().head()

Writer          45
Artist          34
Designer        34
Journalist      33
Entrepreneur    31
Name: speaker_occupation, dtype: int64

In [136]:
ted.speaker_occupation.value_counts().tail()

NASA scientist                                 1
Musician and inventor                          1
Peace activist, Nobelist                       1
Policy entrepreneur, climate expert, author    1
Folk musician                                  1
Name: speaker_occupation, dtype: int64

Because we are not going to be able to effectively clean all the values and separate them .... we might even just drop the lower frequency occupations.....

In [133]:
ted.ratings_list.head()

0    [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
1    [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...
2    [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...
3    [{'id': 3, 'name': 'Courageous', 'count': 760}...
4    [{'id': 9, 'name': 'Ingenious', 'count': 3202}...
Name: ratings_list, dtype: object

In [123]:
def get_funny_count(list_of_dicts, count_list):
    for d in list_of_dicts:
        for k in d:
            if k['name'] == 'Funny':
                count_list.append(k['count'])       
    return count_list

In [132]:
count_list = []
get_funny_count(ted.ratings_list, count_list)[:10]

[19645, 544, 964, 59, 1390, 1102, 2484, 9, 60, 26]

In [125]:
ted['funny_count'] = count_list

In [134]:
ted.speaker_occupation.value_counts().head()

Writer          45
Artist          34
Designer        34
Journalist      33
Entrepreneur    31
Name: speaker_occupation, dtype: int64

In [147]:
ted.groupby('speaker_occupation').funny_count.agg(['count','mean', 'sum']).sort_values(by='mean').tail()

count    mean    sum
speaker_occupation                               
Vulnerability researcher         2  3088.0   6176
Actor, comedian, playwright      2  3254.5   6509
Blogger                          2  3773.0   7546
Comedian and writer              2  4970.5   9941
Author/educator                  4  6190.0  24760

# Step 1: Count the number of funny ratings
We should check to see if funny is always in the list

In [148]:
# We can use the ratings column here and apply str method and because this will return T/F we will use value_counts()
ted.ratings.str.contains('Funny').value_counts()

True    2550
Name: ratings, dtype: int64

So we quickly see that Funny is always going to be in our dictionary, but the value for counts could be 0

In [151]:
def get_funny_ratings(list_of_dicts):
    for d in list_of_dicts:
        if d['name'] == 'Funny':
            return d['count']

In [152]:
ted.ratings_list.apply(get_funny_ratings)

0       19645
1         544
2         964
3          59
4        1390
        ...  
2545        1
2546       20
2547        1
2548       63
2549        0
Name: ratings_list, Length: 2550, dtype: int64

We would normally save this to a column in the dataframe but I have already done it as `funny_count`.

In [153]:
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

   ...                                               tags  \
0  ...  ['children', 'creativity', 'culture', 'dance',...   
1  ...  ['alternative energy', 'cars', 'climate change...   
2  ...  ['computers', 'entertainment', 'interface desi...   
3  ...  ['MacArthur grant', 'activism', 'business', 'c...   
4  ...  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  \
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520   
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292   
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550   
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869   

  comments_per_view views_per_comment  film_datetime  \
0          0.000096      10372.745443     2006-02-25   
1          0.000083      12077.433962     2006-02-25   
2          0.000076      13195.903226     2006-02-24   
3          0.000118       8487.750000     2006-02-26   
4          0.000049      20245.984823     2006-02-22   

                                        ratings_list  num_ratings funny_count  
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...        93850       19645  
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...         2936         544  
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...         2824         964  
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...         3728          59  
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...        25620        1390  

[5 rows x 23 columns]

# Step 2: Calculate the rate of funny ratings for each talk

In [154]:
ted['funny_rate'] = ted.funny_count / ted.num_ratings

In [158]:
ted.sort_values('funny_rate').speaker_occupation.tail(20)

1849                       Science humorist
337                                Comedian
124     Performance poet, multimedia artist
315                                  Expert
1168             Social energy entrepreneur
1468                          Ornithologist
595                  Comedian, voice artist
1534                         Cartoon editor
97                                 Satirist
2297                          Actor, writer
568                                Comedian
675                          Data scientist
21                     Humorist, web artist
194                                Jugglers
2273                    Comedian and writer
2114                    Comedian and writer
173                                Investor
747                                Comedian
1398                               Comedian
685             Actor, comedian, playwright
Name: speaker_occupation, dtype: object

These are the top 20 talks with the highest funny_rate

# Step 3: Analyze the funny rate by occupation

In [160]:
ted.groupby('speaker_occupation').funny_rate.agg(['count','mean','sum']).sort_values(by='mean').tail()

count      mean       sum
speaker_occupation                                    
Comedian                         6  0.512457  3.074740
Actor, writer                    1  0.515152  0.515152
Actor, comedian, playwright      2  0.558107  1.116214
Jugglers                         1  0.566828  0.566828
Comedian and writer              2  0.602085  1.204171

In [161]:
ted.speaker_occupation.describe()

count       2544
unique      1458
top       Writer
freq          45
Name: speaker_occupation, dtype: object

The weakness in our approach is that there are 1458 Unique values meaing that there us only one of them. So a small sample size is our biggest problems here.

# Step 4: Focus on occupation that are well represented in the data.

In [162]:
ted.speaker_occupation.value_counts()

Writer                                         45
Artist                                         34
Designer                                       34
Journalist                                     33
Entrepreneur                                   31
                                               ..
NASA scientist                                  1
Musician and inventor                           1
Peace activist, Nobelist                        1
Policy entrepreneur, climate expert, author     1
Folk musician                                   1
Name: speaker_occupation, Length: 1458, dtype: int64

We want to generate a list of fairly frequent occupations.

What is the dtype of the value_counts output?

In [166]:
type(ted.speaker_occupation.value_counts()[0])

numpy.int64

In [167]:
type(ted.speaker_occupation.value_counts())

pandas.core.series.Series

The value_counts output is a pandas series. So we can manipulateit as if it were a anyother part of the dataframe.

In [168]:
# we can save it and filter it like any other series
occupation_counts = ted.speaker_occupation.value_counts()

In [169]:
occupation_counts[occupation_counts >= 5]

Writer                45
Artist                34
Designer              34
Journalist            33
Entrepreneur          31
                      ..
Physician              5
Chemist                5
Sculptor               5
Sound consultant       5
Musician, activist     5
Name: speaker_occupation, Length: 68, dtype: int64

In [170]:
# Now we can take the index 
occupation_counts[occupation_counts >= 5].index

Index(['Writer', 'Artist', 'Designer', 'Journalist', 'Entrepreneur',
       'Architect', 'Inventor', 'Psychologist', 'Photographer', 'Filmmaker',
       'Economist', 'Educator', 'Neuroscientist', 'Author', 'Roboticist',
       'Philosopher', 'Biologist', 'Physicist', 'Musician', 'Marine biologist',
       'Global health expert; data visionary', 'Technologist', 'Activist',
       'Astronomer', 'Poet', 'Singer/songwriter', 'Graphic designer',
       'Behavioral economist', 'Oceanographer', 'Philanthropist', 'Historian',
       'Engineer', 'Futurist', 'Social psychologist', 'Computer scientist',
       'Novelist', 'Mathematician', 'Astrophysicist', 'Comedian',
       'Performance poet, multimedia artist', 'Legal activist',
       'Social entrepreneur', 'Singer-songwriter', 'Photojournalist',
       'Reporter', 'Techno-illusionist', 'Writer, activist',
       'Evolutionary biologist', 'Climate advocate', 'Researcher',
       'Data scientist', 'Science writer', 'Surgeon', 'Violinist',
     

In [171]:
top_occupations = occupation_counts[occupation_counts >= 5].index

In [172]:
type(top_occupations)

pandas.core.indexes.base.Index

We can see that this is an Index, but an Index can be treated like a list.

In [174]:
ted[ted.speaker_occupation.isin(top_occupations)].head()

comments                                        description  duration  \
1        265  With the same humor and humanity he exuded in ...       977   
4        593  You've never seen data presented like this. Wi...      1190   
7         46  Architect Joshua Prince-Ramus takes the audien...      1198   
11        55  Jehane Noujaim unveils her 2006 TED Prize wish...      1538   
14        99  Nicholas Negroponte, founder of the MIT Media ...      1057   

      event   film_date  languages         main_speaker  \
1   TED2006  1140825600         43              Al Gore   
4   TED2006  1140566400         48         Hans Rosling   
7   TED2006  1140652800         19  Joshua Prince-Ramus   
11  TED2006  1140912000         20       Jehane Noujaim   
14  TED2006  1140652800         25  Nicholas Negroponte   

                                                 name  num_speaker  \
1                Al Gore: Averting the climate crisis            1   
4       Hans Rosling: The best stats you've ever seen            1   
7   Joshua Prince-Ramus: Behind the design of Seat...            1   
11      Jehane Noujaim: My wish: A global day of film            1   
14          Nicholas Negroponte: One Laptop per Child            1   

    published_date  ...                                   title  \
1       1151367060  ...             Averting the climate crisis   
4       1151440680  ...         The best stats you've ever seen   
7       1152490260  ...  Behind the design of Seattle's library   
11      1153786260  ...           My wish: A global day of film   
14      1154391060  ...                    One Laptop per Child   

                                                  url     views  \
1   https://www.ted.com/talks/al_gore_on_averting_...   3200520   
4   https://www.ted.com/talks/hans_rosling_shows_t...  12005869   
7   https://www.ted.com/talks/joshua_prince_ramus_...    967741   
11  https://www.ted.com/talks/jehane_noujaim_inspi...    387877   
14  https://www.ted.com/talks/nicholas_negroponte_...    358304   

   comments_per_view views_per_comment film_datetime  \
1           0.000083      12077.433962    2006-02-25   
4           0.000049      20245.984823    2006-02-22   
7           0.000048      21037.847826    2006-02-23   
11          0.000142       7052.309091    2006-02-26   
14          0.000276       3619.232323    2006-02-23   

                                         ratings_list  num_ratings  \
1   [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...         2936   
4   [{'id': 9, 'name': 'Ingenious', 'count': 3202}...        25620   
7   [{'id': 9, 'name': 'Ingenious', 'count': 195},...         1024   
11  [{'id': 1, 'name': 'Beautiful', 'count': 70}, ...          592   
14  [{'id': 3, 'name': 'Courageous', 'count': 83},...          781   

    funny_count funny_rate  
1           544   0.185286  
4          1390   0.054254  
7             9   0.008789  
11            5   0.008446  
14           14   0.017926  

[5 rows x 24 columns]

In [175]:
ted_top_occupation = ted[ted.speaker_occupation.isin(top_occupations)]

In [176]:
ted_top_occupation.shape

(786, 24)

In [180]:
ted_top_occupation.groupby('speaker_occupation').funny_rate.mean().sort_values().tail()

speaker_occupation
Data scientist                         0.184076
Producer                               0.202531
Singer/songwriter                      0.252205
Performance poet, multimedia artist    0.306468
Comedian                               0.512457
Name: funny_rate, dtype: float64

Some weaknesses....5 is still a small sample size
.....there are still some multi categories that overlap
.....the individual might be funny but it may not be representative of the profession

**Lessons**
1. Check your assumptions about your data
    - did you assume that funny was in all those rows?
2. Check if your results makes sense
3. Take advantage of the fact that pandas operations often output a DataFrame or a Series
4. Watch out for small sample sizes
5. Consider the impact of missing data
6. Data Scientists are hilarious